src/System.Security.Cryptography.Algorithms/tests/HKDFTests.cs

```diff
@@ -0,0 +1,611 @@
+        {
+            byte[] ikm = new byte[20];
+            byte[] salt = new byte[20];
+            Assert.Throws<ArgumentOutOfRangeException>(() => HKDF.Extract(default(HashAlgorithmName), ikm, salt));
+        }
+
+        [Fact]
+        public void Rfc5869ExtractByteArrayNonsensicalHash()
+        {
+            byte[] ikm = new byte[20];
+            byte[] salt = new byte[20];
+            Assert.Throws<ArgumentOutOfRangeException>(() => HKDF.Extract(new HashAlgorithmName("foo"), ikm, salt));
+        }
+
+        [Fact]
+        public void Rfc5869ExtractByteArrayNullIkm()
+        {
+            byte[] salt = new byte[20];
+            Assert.Throws<ArgumentNullException>(() => HKDF.Extract(HashAlgorithmName.SHA1, null, salt));
+        }
+
+        [Fact]
+        public void Rfc5869ExtractByteArrayEmptyIkm()
+        {
+            byte[] salt = new byte[20];
+            byte[] ikm = Array.Empty<byte>();
+
+            // Ensure does not throw
+            byte[] prk = HKDF.Extract(HashAlgorithmName.SHA1, ikm, salt);
+            Assert.Equal("FBDB1D1B18AA6C08324B7D64B71FB76370690E1D".HexToByteArray(), prk);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetRfc5869TestCases))]
+        public void Rfc5869ExtractSpanTests(Rfc5869TestCase test)
+        {
+            byte[] prk = new byte[test.Prk.Length];
+            Assert.Equal(test.Prk.Length, HKDF.Extract(test.Hash, test.Ikm, test.Salt, prk));
+            Assert.Equal(test.Prk, prk);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetRfc5869TestCases))]
+        public void Rfc5869ExtractSpanTamperIkmTests(Rfc5869TestCase test)
+        {
+            test.Ikm[0] ^= 1;
+            byte[] prk = new byte[test.Prk.Length];
+            HKDF.Extract(test.Hash, test.Ikm, test.Salt, prk);
+            Assert.NotEqual(test.Prk, prk);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetRfc5869TestCasesWithNonEmptySalt))]
+        public void Rfc5869ExtractSpanTamperSaltTests(Rfc5869TestCase test)
+        {
+            test.Salt[0] ^= 1;
+            byte[] prk = new byte[test.Prk.Length];
+            HKDF.Extract(test.Hash, test.Ikm, test.Salt, prk);
+            Assert.NotEqual(test.Prk, prk);
+        }
+
+        [Fact]
+        public void Rfc5869ExtractSpanDefaultHash()
+        {
+            byte[] prk = new byte[20];
+            byte[] ikm = new byte[20];
+            byte[] salt = new byte[20];
+            Assert.Throws<ArgumentOutOfRangeException>(() => HKDF.Extract(default(HashAlgorithmName), ikm, salt, prk));
+        }
+
+        [Fact]
+        public void Rfc5869ExtractSpanNonsensicalHash()
+        {
+            byte[] prk = new byte[20];
+            byte[] ikm = new byte[20];
+            byte[] salt = new byte[20];
+            Assert.Throws<ArgumentOutOfRangeException>(() => HKDF.Extract(new HashAlgorithmName("foo"), ikm, salt, prk));
+        }
+
+        [Fact]
+        public void Rfc5869ExtractSpanEmptyIkm()
```

Comment:
If you feel like doing a rewrite of all the tests, you could do something like

```C#
public class HkdfTests
{
    protected abstract int Extract(HashAlgorithmName hash, byte[] ikm, byte[] salt, out byte[] prk);
    // Similar for Expand, DeriveKey

    [Fact]
    public void Rfc5869ExtractEmptyIkm()
    {
            byte[] ikm = Array.Empty<byte>();
            byte[] salt = new byte[20];
            Assert.Equal(20, Extract(HashAlgorithmName.SHA1, ikm, salt, out byte[] prk));
            Assert.Equal("FBDB1D1B18AA6C08324B7D64B71FB76370690E1D", prk.ByteArrayToHex());
    }

    // etc
}

public class HkdfTests_Span : HkdfTests
{
    protected override int Extract(HashAlgorithmName hash, byte[] ikm, byte[] salt, out byte[] prk)
    {
        // Maybe prkSize should be passed in instead of recalculated.  No worries.
        prk = new byte[20];
        return HKDF.Expand(hash, ikm, salt, prk);
    }

    // etc
}

public class HkdfTests_Span : HkdfTests
{
    protected override int Extract(HashAlgorithmName hash, byte[] ikm, byte[] salt, out byte[] prk)
    {
        prk = HKDF.Expand(hash, ikm, salt);
        return prk.Length;
    }

    // etc
}
```


Then you only need to write each test once.  (It's a pattern we've done a bit with simplifying span/array overload testing)

The tests that are unique to each side (like null input arrays throw, and incorrectly-sized span for PRK) would then go in the respective derived types.

This is also cleanup that you could do post-merge.